solutions/ai-chatgpt/components/ChatBrowser.tsx

```diff
@@ -0,0 +1,137 @@
+import { useEffect, useState } from 'react'
+import { useCookies } from 'react-cookie'
+
+import { type Message, ChatLine, LoadingChatLine } from './ChatLine'
+import { InputMessage } from './InputMessage'
+import InputKey from './InputKey'
+import { generatePromptFromMessages } from '../lib/generatePromptFromMessages'
+
+const COOKIE_NAME = 'nextjs-example-ai-chat-gpt3-user'
+const COOKIE_KEY = 'nextjs-example-ai-chat-gpt3-key'
+
+// default first message to display in UI (not necessary to define the prompt)
+export const initialMessages: Message[] = [
+  {
+    who: 'bot',
+    message: 'Hi! I’m A friendly AI assistant. Ask me anything!',
+  },
+]
+
+export function Chat() {
+  const [messages, setMessages] = useState<Message[]>(initialMessages)
+  const [input, setInput] = useState('')
+  const [loading, setLoading] = useState(false)
+  const [cookie, setCookie] = useCookies([COOKIE_NAME, COOKIE_KEY])
+
+  const [botName, setBotName] = useState('AI')
+  const [userName, setUserName] = useState('News reporter')
+  const [keyOpenAI, setKeyOpenAI] = useState('sk-XXXX') // default OpenAI key
+  const [prompt, setPrompt] = useState(
+    `I am Friendly AI Assistant. \n\nThis is the conversation between AI Bot and a news reporter.\n\n${botName}: ${initialMessages[0].message}\n${userName}: `
+  )
+
+  useEffect(() => {
+    if (!cookie[COOKIE_NAME]) {
+      // generate a semi random short id
+      const randomId = Math.random().toString(36).substring(7)
+      setCookie(COOKIE_NAME, randomId)
+    }
+    // override the default key if it's set in the cookie
+    if (cookie[COOKIE_KEY]) {
+      setKeyOpenAI(cookie[COOKIE_KEY])
+    }
+  }, [cookie, setCookie])
+
+  const sendMessage = async (message: string) => {
+    setLoading(true)
+    const newMessages = [
+      ...messages,
+      { message: message, who: 'user' } as Message,
+    ]
+    setMessages(newMessages)
+    const last10messages = newMessages.slice(-12) // remember last 12 messages
+
+    // const messages = req.body.messages
+    const messagesPrompt = generatePromptFromMessages(
+      last10messages,
+      userName,
+      botName
+    )
+    const finalPrompt = `${prompt}${messagesPrompt}\n${botName}: `
+
+    const payload = {
+      model: 'text-davinci-003',
+      prompt: finalPrompt,
+      temperature: process.env.AI_TEMP ? parseFloat(process.env.AI_TEMP) : 0.7,
+      max_tokens: process.env.AI_MAX_TOKENS
+        ? parseInt(process.env.AI_MAX_TOKENS)
+        : 200,
+      top_p: 1,
+      frequency_penalty: 0,
+      presence_penalty: 0,
+      stop: [`${botName}:`, `${userName}:`],
+      user: cookie[COOKIE_NAME],
+    }
+
+    const requestHeaders: Record<string, string> = {
+      'Content-Type': 'application/json',
+      Authorization: `Bearer ${keyOpenAI}`,
+    }
+
+    // use OpenAI API directly in browser to generate a response
```

Comment:
This makes me nervous 😅

Reply:
This BYOK idea really appeals to me. I have seen many examples of this. In this particular example, even if the user accidentally leaks the OpenAI key since OpenAI has soft and hard limits, damage is somewhat limited. 
Idk. Maybe you're right. How do others accomplish BYOK?

Reply:
Let's not recommend sharing API keys publicly unless OpenAI explicitly says somewhere in their docs that it's okay to do so.